### ISSUE TITLE
Sorting Json

### ISSUE THREAD

**Issue by vincentytran** [schibsted/jslt#172]
I am wondering if there is a function that be used to sort json. Something that could sort an array using a specified value from each element.

e.g. you can sort by the score field in the following example:
[ { "id": "a", "score": 4}, { "id": "b", "score": 3}]

**Comment by vincentytran** [comment]
I figured out a somewhat "work around" to sort using the group-by which seems to always be ascending order. To get something sorted in reverse order you can subtract the number from a maximum value.

input { "data": [ { "id": "a", "score": "4"}, { "id": "b", "score": "3"}] }

import "http://jslt.schibsted.com/2018/experimental" as exp
let sorted = [for(exp:group-by(.data, 10 - number(.score), .)) .values[0] ]
$sorted

[ {
  "id" : "a",
  "score" : "4"
}, {
  "id" : "b",
  "score" : "3"
} ]

**Comment by larsga** [comment]
I've been thinking about adding this one. It would have to be a macro, but that's doable. Then you could sort that list by doing

```
sort(. , .score)
```

Maybe adding `true` as a third parameter to get descending order.

The challenge is that this requires adding some sophistication to the data model of JSLT, because now we need a notion of order for each data type, as well as ordering across data types. But it definitely can be done, and I think probably it should be. I was kind of waiting for this one to show up.

**Comment by hemakumarg89** [comment]
Hi, Is the Sorting function implemented in JSLT as discussed above?

**Comment by larsga** [comment]
No. Work this year has so far focused on removing the Jackson dependency (issue #123), but it looks like this is something we need to give priority soon.

**Comment by larsga** [comment]
I think the only practical way to do it at the moment is to write a function in Java and plug it in. Unfortunately.

**Comment by hemakumarg89** [comment]
Thanks for the response. I will look in to it.

**Comment by hemakumarg89** [comment]
> No. Work this year has so far focused on removing the Jackson dependency (issue #123), but it looks like this is something we need to give priority soon.

Hi @larsga , Any plans to implement sorting in JSLT anytime soon ? or any other functions recently added through which the sorting could be achieved please ? Thanks in Advance

**Comment by larsga** [comment]
There's another use case on [StackOverflow](https://stackoverflow.com/questions/76236450/jslt-are-there-global-variables)

**Comment by abhishekrongali** [comment]
Was above sorting implemented? Can you please help with the syntax?

**Comment by larsga** [comment]
Sorting has not been implemented, but I think it should be.

To implement it we will need a design document, since this is non-trivial functionality with some subtleties to it. I'll see if I can get started on that.

**Comment by larsga** [comment]
@abhishekrongali and others who want sorting: it would be very helpful if you could give concrete examples of problems you want to solve with sorting. That would help me evaluate design choices.

**Comment by larsga** [comment]
I wrote up [a design document](https://github.com/schibsted/jslt/blob/master/docs/sorting.md). Comments in this thread very much welcome.

**Comment by abhishekrongali** [comment]
What would be the syntax to sort below example?
I want to sort the json request containing array's based name and update the json request in sorted format
Before sorting -> { "students": [ { "id": "a", "name": "roger"}, { "id": "b", "name": "drake"}] }
After sorting expectation ->  { "students": [ { "id": "b", "name": "drake"},{ "id": "a", "name": "roger"}] }

**Comment by larsga** [comment]
The example doesn't look 100% complete, but I assume you'd do something like the below, if you wanted to return the same input except you'd sort the `students` key:

```
{
  "students" : .students order by .name,
  * : .
}
```

Or, alternatively:

```
{
  "students" : sort(.students, .name),
  * : .
}
```

**Comment by abhishekrongali** [comment]
Thanks for above explanation! We are using jslt transformation in apache camel! Functions sort() or order-by() is not working in jslt! Throws error as no such function..  Could you please help here? I wanted to sort an array while looping it with for loop..

**Comment by larsga** [comment]
I need to implement it before it becomes available, I'm afraid.

It is possible to implement it yourself as a function inside JSLT, however, but it won't be fast. The easiest to implement is probably insertion sort, which is O(n^2). Mergesort and quicksort will probably create too many temporary arrays and spend too much time copying data. At least with larger arrays.
